src/error.rs

```diff
@@ -129,6 +137,12 @@ pub(crate) enum Error {
   SymlinkRoot { root: PathBuf },
   #[snafu(display("Failed to retrieve system time: {}", source))]
   SystemTime { source: SystemTimeError },
+  #[snafu(display("The UDP tracker connection id has expired."))]
+  UdpConnectionIdExpired,
+  #[snafu(display("UDP tracker has failed to respond 9 times."))]
+  UdpConnectionExhaustedRetries,
+  #[snafu(display("UDP tracker response differed from expectations"))]
+  UdpTrackerBadResponse,
```

Comment:
The `UdpConnectionIdExpired` error is presently returned when a client with stale `connection_id` tries to send a message. Here's the relevant section of BEP15: 

> UDP connections / spoofing
> In the ideal case, only 2 packets would be necessary. However, it is possible to spoof the source address of a UDP packet. The tracker has to ensure this doesn't occur, so it calculates a value (connection_id) and sends it to the client. If the client spoofed it's source address, it won't receive this value (unless it's sniffing the network). The connection_id will then be send to the tracker again in packet 3. The tracker verifies the connection_id and ignores the request if it doesn't match. Connection IDs should not be guessable by the client. This is comparable to a TCP handshake and a syn cookie like approach can be used to storing the connection IDs on the tracker side. A connection ID can be used for multiple requests. A client can use a connection ID until one minute after it has received it. Trackers should accept the connection ID until two minutes after it has been send.

Reply:
This is only used in `udp_tracker.rs:send_and_retry_with_backoff()`. The timings look something like: 

    /\ connection id lasts 1 minute
    /\ for n \in [0,8], back off for 15 * 2**n seconds before trying to resend

Upshot: The connection id will need to be refreshed at some point during the retrying. And but refreshing `connection_id` requires sending another message.

So I think there's some potential for live locking (?) that is sidestepped by completely erroring out once the `connection_id` is found to be stale.

Reply:
That all makes sense. In that case I think `TrackerConnectionExpired` is a better name for the error, since conceptually it's the whole connection that's expiring, although that's tracked by the id.